pkg/frame.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+var compoundTypeRegEx = regexp.MustCompile(`(Array|Tuple)\(([A-Za-z0-9,() ]+)\)`)
+
+type ClickHouseFrame struct {
+	RefId  string
+	Name   string
+	Fields []*ClickHouseField
+}
+
+func NewFrame(refId string, name string, fieldsMeta []*FieldMeta, tz FetchTZ) *ClickHouseFrame {
+	fields := make([]*ClickHouseField, len(fieldsMeta))
+
+	for i, meta := range fieldsMeta {
+		fields[i] = NewField(meta.Name, meta.Type, tz)
+
+		field := fields[i]
+		fieldType := meta.Type
+		compoundName := meta.Name
+		for compoundTypeRegEx.MatchString(fieldType) {
+			compoundMeta := compoundTypeRegEx.FindStringSubmatch(fieldType)
+			compoundType := compoundMeta[1]
+			fieldType = compoundMeta[2]
+
+			compoundName += "-" + compoundType
+
+			if field.Fields != nil {
+				field = field.Fields[0]
+			}
+
+			field.IsCompound = true
+			field.Fields = make([]*ClickHouseField, 1)
+			field.Fields[0] = NewField(compoundName, compoundType, tz)
+		}
+
+		// Parse the last compound item
+		if field.IsCompound {
+			compoundMeta := strings.Split(fieldType, ",")
+			compoundParts := make([]*ClickHouseField, len(compoundMeta))
+			for i := 0; i < len(compoundMeta); i++ {
+				compoundType := strings.TrimSpace(compoundMeta[i])
+				compoundParts[i] = NewField(compoundName+"-"+compoundType, compoundType, tz)
```

Comment:
when we get compound type after apply $macros, like Array(Tupple()
string element of Tuple is a Time Series Name, numeric element of Tuple is a series value
please look to https://github.com/Vertamedia/clickhouse-grafana/blob/v2.2.3/datasource.go#L287-L323 for an explanation